src/store_client.cc

```diff
@@ -702,13 +694,12 @@ storeUnregister(store_client * sc, StoreEntry * e, void *data)
         return 0;
     }
 
-    if (mem->clientCount() == 0) {
+    if (mem->clients.size() == 0) {
         debugs(90, 3, "storeUnregister: Consistency failure - store client being unregistered is not in the mem object's list for '" << e->getMD5Text() << "'");
         return 0;
     }
 
-    dlinkDelete(&sc->node, &mem->clients);
-    -- mem->nclients;
+    mem->clients.remove(sc);
```

Comment:
Huh? that pre-C++20 check is *exactly* what the if-statement directly above the remove() call is doing already.

I considered using find() to be more accurate, but that is a huge performance impact just to consistency check a situation which does not seem to have much in the way of side effects.

Reply:
> that pre-C++20 check is exactly what the if-statement directly above the remove() call is doing already.

No, that `if` statement checks that we are not removing from an empty list. We should (also) be checking that we are removing from the given StoreEntry client list rather than from some other StoreEntry client list. For example, given two _different_ StoreEntry objects, each with its own non-empty client list, and the _same_ store_client pointer, the method should declare a BUG in at least one of those two calls.

Reply:
> I considered using find() to be more accurate, but that is a huge performance impact

Actually, the opposite is true:

* In our use case, on average, std::remove() is twice as expensive as find()+erase() because the former always scans the entire list while the latter stops at the first found element (scanning just half of the list, on average).

* In a more general case that includes duplicate entries (which we do not have to deal with), both approaches have about the same performance because they have to scan the entire list.

Reply:
Doh!, of course. I was thinking find()+remove() efficiency. Will do this change

Reply:
Latest code does this.